### PR TITLE
add test_auto_lstm for Autolstm

### PR DESCRIPTION
**Pull request by liangs6212**

### CODE REVIEW

File: pyzoo/zoo/chronos/examples/lstm/test_auto_lstm.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+                      shuffle=True)
+
+
+def valid_dataloader_creator(config):
+    return DataLoader(RandomDataset(data=val_data),
+                      batch_size=config["batch_size"],
+                      shuffle=True)
+
+
+def test_fit_np(args):
```
Comment (yushan111) [inline pyzoo/zoo/chronos/examples/lstm/test_auto_lstm.py:84]: change name

File: pyzoo/zoo/chronos/examples/lstm/test_auto_lstm.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+    def __init__(self, data):
+        x, y = get_x_y(data)
+        self.x = torch.from_numpy(x).float()
+        self.y = torch.from_numpy(y).float()
+
+    def __len__(self):
+        return self.x.shape[0]
+
+    def __getitem__(self, idx):
```
Comment (yushan111) [inline pyzoo/zoo/chronos/examples/lstm/test_auto_lstm.py:68]: move the rolling process in `get_x_y` here

File: pyzoo/zoo/chronos/examples/lstm/test_auto_lstm.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+def get_x_y(data):
+    X, y = [], []
+    for i in range(len(data) - past_seq_len - future_seq_len - 1):
+        X.append(data[i: i + past_seq_len, 0: input_feature_dim])
+        y.append(data[i + past_seq_len: i +
+                 past_seq_len + future_seq_len, -output_feature_dim:])
+    return np.array(X), np.array(y)
+
+
+class RandomDataset(Dataset):
```
Comment (yushan111) [inline pyzoo/zoo/chronos/examples/lstm/test_auto_lstm.py:58]: change name

**Comment by liangs6212** [comment]
http://10.239.176.111:18888/job/ZOO-PR-Python-ExampleTests-py37-Spark3/1231/
http://10.239.176.111:18888/job/ZOO-PR-Validation/5761/

File: pyzoo/zoo/chronos/examples/lstm/README.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+```
+
+## Options
+* `--cluster_mode` The mode for the Spark cluster. local or yarn. Default to be `local`. You can refer to OrcaContext documents [here](https://analytics-zoo.readthedocs.io/en/latest/doc/Orca/Overview/orca-context.html) for details.
+* `--num_workers` The number of workers to be used in the cluster. You can change it depending on your own cluster setting. Default to be 2.
+* `--cores` "The number of cpu cores you want to use on each node. Default to be 4.
+* `--memory` The memory you want to use on each node. Default to be 10g.
+* `--backend` The backend of the lstm model. We only support backend as 'torch' for now.
+* `--epoch` Max number of epochs to train in each trial. Default to be 1.
+* `--cpus_per_train` Number of cpus for each trial. Default to be 2.
```
Comment (yushan111) [inline pyzoo/zoo/chronos/examples/lstm/README.md:34]: trial

File: pyzoo/zoo/chronos/examples/lstm/README.md [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+```
+python test_auto_lstm.py --cluster_model yarn
+```
+
+## Options
+* `--cluster_mode` The mode for the Spark cluster. local or yarn. Default to be `local`. You can refer to OrcaContext documents [here](https://analytics-zoo.readthedocs.io/en/latest/doc/Orca/Overview/orca-context.html) for details.
+* `--num_workers` The number of workers to be used in the cluster. You can change it depending on your own cluster setting. Default to be 2.
+* `--cores` "The number of cpu cores you want to use on each node. Default to be 4.
+* `--memory` The memory you want to use on each node. Default to be 10g.
+* `--backend` The backend of the lstm model. We only support backend as 'torch' for now.
```
Comment (yushan111) [inline pyzoo/zoo/chronos/examples/lstm/README.md:32]: remove

File: pyzoo/zoo/chronos/examples/lstm/autolstm_household_electricity.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+import pandas as pd
+import argparse
+
+
+from zoo.chronos.autots.model.auto_lstm import AutoLSTM
+from zoo.orca.automl import hp
+from zoo.orca import init_orca_context, stop_orca_context
+
+input_feature_dim = 7
+output_feature_dim = 2
```
Comment (yushan111) [inline pyzoo/zoo/chronos/examples/lstm/autolstm_household_electricity.py:34]: move to main

File: pyzoo/zoo/chronos/examples/lstm/autolstm_household_electricity.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+                        help="Number of times to sample from the search_space.")
+
+    args = parser.parse_args()
+
+    num_nodes = 1 if args.cluster_mode == "local" else args.num_workers
+    init_orca_context(cluster_mode=args.cluster_mode, cores=args.cores,
+                      memory=args.memory, num_nodes=num_nodes, init_ray_on_spark=True)
+
+    auto_lstm = AutoLSTM(input_feature_num=input_feature_dim,
+                         output_target_num=output_feature_dim,
```
Comment (yushan111) [inline pyzoo/zoo/chronos/examples/lstm/autolstm_household_electricity.py:115]: input_feature_num = 7
output_feature_num = 2

File: pyzoo/zoo/chronos/examples/lstm/autolstm_household_electricity.py [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+
+
+from zoo.chronos.autots.model.auto_lstm import AutoLSTM
+from zoo.orca.automl import hp
+from zoo.orca import init_orca_context, stop_orca_context
+
+input_feature_dim = 7
+output_feature_dim = 2
+past_seq_len = 9
+future_seq_len = 1
```
Comment (yushan111) [inline pyzoo/zoo/chronos/examples/lstm/autolstm_household_electricity.py:36]: move to dataset, self....

**Review by yushan111 — COMMENTED** [review]
We may also need to enable auto test for the example

**Comment by liangs6212** [comment]
http://10.239.176.111:18888/job/ZOO-PR-Validation-Spark-2.1/5775/
http://10.239.176.111:18888/job/ZOO-PR-Validation-Other/5793/
http://10.239.176.111:18888/job/ZOO-PR-Validation/5768/
http://10.239.176.111:18888/job/ZOO-PR-Pip-ExampleTests-py36/31/

File: pyzoo/zoo/chronos/examples/lstm/autolstm_household_electricity.py [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+                         name="auto_lstm")
+    auto_lstm.fit(data=train_dataloader_creator,
+                  epochs=args.epoch,
+                  batch_size=hp.choice([32, 64]),
+                  validation_data=valid_dataloader_creator,
+                  n_sampling=args.n_sampling,
+                  )
+    best_model = auto_lstm.get_best_model()
+    best_model_rmse = best_model._validate(
+        valid_dataloader_creator(best_model.config), metric='rmse')
```
Comment (yushan111) [inline pyzoo/zoo/chronos/examples/lstm/autolstm_household_electricity.py:132]: We shouldn't use private method, use `evaluate` instead.

File: pyzoo/zoo/chronos/examples/lstm/autolstm_household_electricity.py [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+    print(f'Evaluation result is {best_model_rmse["rmse"]:.2f}')
```
Comment (yushan111) [inline pyzoo/zoo/chronos/examples/lstm/autolstm_household_electricity.py:133]: Please add the evaluation result in comment

**Review by yushan111 — COMMENTED** [review]
Others LGTM

**Comment by liangs6212** [comment]
mse is 0.003540067975989188, smape is 6.76483013674847

**Comment by yushan111** [comment]
We may also change the data processing part to TSDataset interface

**Comment by liangs6212** [comment]
mse is 0.13392152918528033, smape is 5.727508936773413

File: pyzoo/zoo/chronos/examples/lstm/autolstm_household_electricity.py [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+                                           horizon=future_seq_len)\
+                                     .to_numpy()
+        test_data = TSDataset.from_pandas(df[int(len(df)*0.8):], dt_col="datetime",
+                                          target_col=["Global_active_power"],
+                                          extra_feature_col=["Global_reactive_power"])
+        x_test, y_test = test_data.impute(mode="linear")\
+                                  .scale(mm, fit=False)\
+                                  .roll(lookback=past_seq_len,
+                                        horizon=future_seq_len)\
+                                  .to_numpy()
```
Comment (TheaperDeng) [inline pyzoo/zoo/chronos/examples/lstm/autolstm_household_electricity.py:75]: you can simplify the code by this method
```python
for tsdata in [train_data, test_data]:
    tsdataset.operations(fit=tsdata is train_data)
...
```

File: pyzoo/zoo/chronos/examples/lstm/autolstm_household_electricity.py [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+
+
+def train_dataloader_creator(config):
+    return DataLoader(AutoLSTMDataset('train', **config),
+                      batch_size=config["batch_size"],
+                      shuffle=True)
+
+
+def valid_dataloader_creator(config):
+    return DataLoader(AutoLSTMDataset('valid', **config),
```
Comment (yushan111) [inline pyzoo/zoo/chronos/examples/lstm/autolstm_household_electricity.py:90]: `**config` might be a little implicit... You should explicit the parameters you want to pass:  `past_seq_len` and `future_seq_len`

Comment (yushan111) [reply pyzoo/zoo/chronos/examples/lstm/autolstm_household_electricity.py:90]: BTW, what value is the `future_seq_len`?

File: pyzoo/zoo/chronos/examples/lstm/autolstm_household_electricity.py [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+    auto_lstm.fit(data=train_dataloader_creator,
+                  epochs=args.epoch,
+                  batch_size=hp.choice([32, 64]),
+                  validation_data=valid_dataloader_creator,
+                  n_sampling=args.n_sampling,
+                  )
+    best_model = auto_lstm.get_best_model()
+
+    _, _, test_data = AutoLSTMDataset.preprocess_data(
+        'valid', **best_model.config)
```
Comment (yushan111) [inline pyzoo/zoo/chronos/examples/lstm/autolstm_household_electricity.py:146]: Explicate the parameters for dataset

File: pyzoo/zoo/chronos/examples/lstm/autolstm_household_electricity.py [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+            data_selection, past_seq_len, future_seq_len)
+
+    def __len__(self):
+        return self.x.shape[0]
+
+    def __getitem__(self, idx):
+        return self.x[idx], self.y[idx]
+
+    @staticmethod
+    def preprocess_data(data_selection, past_seq_len, future_seq_len, **kwargs):
```
Comment (yushan111) [inline pyzoo/zoo/chronos/examples/lstm/autolstm_household_electricity.py:48]: Shall we move this method out of the dataset class?

Comment (yushan111) [reply pyzoo/zoo/chronos/examples/lstm/autolstm_household_electricity.py:48]: split this into 
`def get_ts_df():`
`def get_tsdataset():`
`def get_data_creator():`
like https://github.com/intel-analytics/analytics-zoo/blob/master/pyzoo/test/zoo/chronos/autots/test_experimental.py

File: pyzoo/zoo/chronos/examples/lstm/autolstm_household_electricity.py [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+                  batch_size=hp.choice([32, 64]),
+                  validation_data=valid_dataloader_creator,
+                  n_sampling=args.n_sampling,
+                  )
+    best_model = auto_lstm.get_best_model()
+
+    _, _, test_data = AutoLSTMDataset.preprocess_data(
+        'valid', **best_model.config)
+    test_data_x, test_data_y = next(
+        iter(valid_dataloader_creator(best_model.config)))
```
Comment (yushan111) [inline pyzoo/zoo/chronos/examples/lstm/autolstm_household_electricity.py:148]: directly use the returned tsdataset in get_tsdataset()？

File: pyzoo/zoo/chronos/examples/lstm/autolstm_household_electricity.py [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+
+    testdata_x_unscale, testdata_y_unscale = test_data._unscale_numpy(
+        test_data_x.numpy()), test_data._unscale_numpy(test_data_y.numpy())
+
+    mse, smape = best_model.evaluate(
+        x=testdata_x_unscale, y=testdata_y_unscale, metrics=['mse', 'smape'])
+    print(f'mse is {np.mean(mse)}, smape is {np.mean(smape)}')
+    print(
```
Comment (yushan111) [inline pyzoo/zoo/chronos/examples/lstm/autolstm_household_electricity.py:156]: best_config = auto_lstm.get_best_config()
print(f"{doc}{best_config})

File: pyzoo/zoo/chronos/examples/lstm/autolstm_household_electricity.py (hunk 12 above)
Comment (liangs6212) [reply pyzoo/zoo/chronos/examples/lstm/autolstm_household_electricity.py:90]: The value of future_seq_len comes from the default value of 1 in search_space.

File: pyzoo/zoo/chronos/examples/lstm/autolstm_household_electricity.py (hunk 16 above)
Comment (liangs6212) [reply pyzoo/zoo/chronos/examples/lstm/autolstm_household_electricity.py:156]: After I updated the version and code, it seems that there is no such get_best_config method

**Comment by liangs6212** [comment]
rmse is 1276.7690209104862, smape is 6.287523931067768

File: pyzoo/zoo/chronos/examples/lstm/autolstm_nyc_taxi.py [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+                                 with_split=True,
+                                 val_ratio=0.1,
+                                 test_ratio=0.1)
+
+
+tsdata_train, tsdata_valid, tsdata_test = get_data_tsdataset()
+stand_scaler = StandardScaler()
+for tsdata in [tsdata_train, tsdata_valid, tsdata_test]:
+    tsdata.impute(mode="linear")\
+          .scale(stand_scaler, fit=(tsdata is tsdata_train))
```
Comment (yushan111) [inline pyzoo/zoo/chronos/examples/lstm/autolstm_nyc_taxi.py:52]: move into `get_data_tsdataset`

**Comment by liangs6212** [comment]
http://10.239.176.111:18888/job/ZOO-PR-Python-ExampleTests/3209/
http://10.239.176.111:18888/job/ZOO-PR-Validation/5943/
http://10.239.176.111:18888/job/ZOO-PR-Python-ExampleTests/3209/

File: pyzoo/zoo/chronos/examples/lstm/autolstm_nyc_taxi.py [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+              horizon=best_config["future_seq_len"])\
+        .to_numpy()
+    yhat = best_model.predict(x)
+
+    y_unscale = tsdata_test.unscale_numpy(y)
+    yhat_unscale = tsdata_test.unscale_numpy(np.expand_dims(yhat, axis=1))
+
+    result = [Evaluator.evaluate(m, y_true=y_unscale, y_pred=yhat_unscale) for m in [
+        'rmse', 'smape']]
+    print(f'rmse is {np.mean(result[0])}, smape is {np.mean(result[1])}')
```
Comment (yushan111) [inline pyzoo/zoo/chronos/examples/lstm/autolstm_nyc_taxi.py:132]: directory use multioutput="uniform_average" in `Evaluator.evaluate`?

Comment (liangs6212) [reply pyzoo/zoo/chronos/examples/lstm/autolstm_nyc_taxi.py:132]: default value "row_values"

Comment (yushan111) [reply pyzoo/zoo/chronos/examples/lstm/autolstm_nyc_taxi.py]: So could that work?
```python
result = [Evaluator.evaluate(m, y_true=y_unscale, y_pred=yhat_unscale, multioutput="uniform_average") for m in ['rmse', 'smape']]
```

Comment (liangs6212) [reply pyzoo/zoo/chronos/examples/lstm/autolstm_nyc_taxi.py]: Okay, I'll change it again, but in fact, this modification is included in the submission of --force.

**Comment by liangs6212** [comment]
http://10.239.176.111:18888/job/ZOO-PR-Pip-ExampleTests-py36/40/
http://10.239.176.111:18888/job/ZOO-PR-Validation/6006/
http://10.239.176.111:18888/job/ZOO-PR-Python-Spark-2.4-py37-ray/2271/
http://10.239.176.111:18888/job/ZOO-PR-Python-ExampleTests-py37-ray-Spark3/1489/
@yushan111
